src/client/auth/aws.rs

```diff
@@ -0,0 +1,414 @@
+
+        let found_access_key = access_key.is_some();
+        let found_secret_key = secret_key.is_some();
+
+        // If we have an access key and secret key, we can continue with the credentials we've
+        // found.
+        if let (Some(access_key), Some(secret_key)) = (access_key, secret_key) {
+            return Ok(Self {
+                access_key,
+                secret_key,
+                session_token,
+            });
+        }
+
+        if found_access_key || found_secret_key {
+            return Err(Error::authentication_error(
+                "MONGODB-AWS",
+                "cannot specify only one of access key and secret key; either both or neither \
+                 must be provided",
+            ));
+        }
+
+        if session_token.is_some() {
+            return Err(Error::authentication_error(
+                "MONGODB-AWS",
+                "cannot specify session token without both access key and secret key",
+            ));
+        }
+
+        if let Ok(relative_uri) = std::env::var("AWS_CONTAINER_CREDENTIALS_RELATIVE_URI") {
+            Self::get_from_ecs(relative_uri, http_client).await
+        } else {
+            Self::get_from_ec2(http_client).await
+        }
+    }
+
+    /// Obtains credentials from the ECS endpoint.
+    async fn get_from_ecs(relative_uri: String, http_client: &HttpClient) -> Result<Self> {
+        // Use the local IP address that AWS uses for ECS agents.
+        let uri = format!("http://169.254.170.2/{}", relative_uri);
+
+        http_client.get_and_deserialize_json(&uri, None).await
+    }
+
+    /// Obtains temporary credentials for an EC2 instance to use for authentication.
+    async fn get_from_ec2(http_client: &HttpClient) -> Result<Self> {
+        let temporary_token = http_client
+            .put_and_read_string(
+                "http://169.254.169.254/latest/api/token",
+                &[("X-aws-ec2-metadata-token-ttl-seconds", "30")],
+            )
+            .await?;
+
+        let role_name_uri = "http://169.254.169.254/latest/meta-data/iam/security-credentials/";
+
+        let role_name = http_client
+            .get_and_read_string(
+                role_name_uri,
+                &[("X-aws-ec2-metadata-token", &temporary_token[..])],
+            )
+            .await?;
+
+        let credential_uri = format!("{}/{}", role_name_uri, role_name);
+
+        http_client
+            .get_and_deserialize_json(
+                &credential_uri,
+                &[("X-aws-ec2-metadata-token", &temporary_token[..])],
+            )
+            .await
+    }
+
+    /// Computes the signed authorization header for the credentials to send to the server in a sasl
+    /// payload.
+    fn compute_authorization_header(
+        &self,
+        date: DateTime<Utc>,
+        host: &str,
+        server_nonce: &[u8],
+    ) -> Result<String> {
+        let date_str = date.format("%Y%m%dT%H%M%SZ").to_string();
```

Comment:
similarly, since this format is used in a couple places, I think a constant for the format string or a function that produces the formatted date would be useful.